plugin/core/sessions.py

```diff
@@ -1898,10 +1910,11 @@ def deduce_payload(
                 return res
         elif "id" in payload:
             response_id = int(payload["id"])
-            handler, result, is_error = self.response_handler(response_id, payload)
-            response_tuple = (handler, result, None, None, None)
-            self._logger.incoming_response(response_id, result, is_error)
-            return response_tuple
+            handler, method, result, is_error = self.response_handler(response_id, payload)
+            self._logger.incoming_response(response_id, deepcopy(result), is_error)
+            if self._plugin and not is_error:
+                self._plugin.on_server_response(method, result)  # type: ignore
```

Comment:
The result could be a primitive like a string or number in which case it won't be possible to modify it when passed through an argument like that. It should be either wrapped into some object or returned through the return value.

Reply:
Ah, good point. I think it's probably better to keep the return value of this function to be `None`, because otherwise plugins need to remember to add something like `else: return result` at the bottom if they choose to implement this method.

I could wrap it in a dict `{"result": result}` and rename the argument to `response`, then plugins could modify `response["result"]`. Or I could use the Response class, then plugins could modify the `response.result` field.
I guess I'll go with the latter option, if there are no objections.

Reply:
Should be fixed by 16024e3e122463c22b0324de594f933aa9c856cf

Reply:
By the way, the type annotation for the result field in both our Response class, as well as in the LSP specs is wrong. The LSP specs misses `array`, and we miss `str`, `float`, `bool`:
https://github.com/sublimelsp/LSP/blob/1289fe9f3a522deac7b554a78b73ba2b8cbe095e/plugin/core/protocol.py#L486

Reply:
Spec lists an `object` which in Javascript means almost any non-literal type so not sure. Depends how it's meant to be interpreted.

Reply:
Hm, just below that, the specs list `data?: string | number | boolean | array | object | null;` for the ResponseError type. So the missing `array` for the type of `result` just looks like an oversight to me. I would guess the `object` refers to a JSON "object" here, i.e. what in Python would be a "dict".

In general there are many inaccuracies and inconsistencies in the LSP specs. Another strange example is that the id of a request is `id: integer | string;`, but for a response message it can also be null: `id: integer | string | null;`.
What should the client do if it receives a response with id `null`? It can't really assign it to any request without having an id...